contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -0,0 +1,406 @@
+extension point returns the answer to "less(pod1, pod2)". Multiple plugins may
+be registered at this point. Plugins registered at this point are called in
+order and the invocation continues as long as plugins return "equal". Once a
+plugin returns "greater" or "smaller" the invocation of these plugins are
+stopped.
+
+### Pre-filter
+
+These plugins are generally useful to check certain conditions that the cluster
+or the Pod must meet. These are also useful to perform pre-processing on the pod
+and store some information about the pod that can be used by other plugins.  
+The pod pointer is passed as an argument to these plugins. If any of these
+plugins return an error, the scheduling cycle is aborted.  
+These plugins are called serially in the same order registered.
+
+### Filter
+
+Filter plugins filter out nodes that cannot run the Pod. Scheduler runs these
+plugins per node in the same order that they are registered, but scheduler may
+run these filter function for multiple nodes in parallel. So, these plugins must
+use synchronization when they modify state.  
+Scheduler stops running the remaining filter functions for a node once one of
+these filters fails for the node.
+
+### Post-filter
+
+The Pod and the set of nodes that can run the Pod are passed to these plugins.
+They are called whether Pod is schedulable or not (whether the set of nodes is
+empty or non-empty).  
+If any of these plugins return an error or if the Pod is determined
+unschedulable, the scheduling cycle is aborted.  
+These plugins are called serially.
+
+### Scoring
+
+These plugins are similar to priority function in scheduler v1. They are
+utilized to rank nodes that have passed the filtering stage. Similar to Filter
+plugins, these are called per node serially in the same order registered, but
+scheduler may run them for multiple nodes in parallel.  
+Each one of these functions return a score for the given node. The score is
+multiplied by the weight of the function and aggregated with the result of other
+scoring functions to yield a total score for the node.  
+These functions can never block scheduling. In case of an error they should
+return zero for the Node being ranked.
+
+### Post-scoring/pre-reservation
+
+After all scoring plugins are invoked and the score of nodes are determined, the
+framework picks the best node with the highest score and then it calls
+post-scoring plugins. The Pod and the chosen Node are passed to these plugins.
+These plugins have one more chance to check any conditions about the assignment
+of the Pod to this Node and reject the node if needed.
+
+![image](images/scheduling-framework-threads.png)
+
+### Reserve
+
+This is not a plugin point. At this point scheduler updates its cache by
+"reserving" a Node (partially or fully) for the Pod. In scheduler v1 this stage
+is called "assume". At this point, only the scheduler cache is updated to
+reflect that the Node is (partially) reserved for the Pod. The actual assignment
```

Comment:
Other plugins may need to do their own reserve as well.  For example, for volumes, we want to mark PV objects as taken in the cache so that we don't choose those PVs again for other pods.

I suppose this could be done at the Admit phase in the plugin, but because Admit is done in a separate thread than the Filter/Scoring/Reserve, then you could have more rejections.

Reply:
Why does the plugin need to modify the scheduler cache? It can keep a separate data structure in its own package to track resources, etc. for its use.  In your case, the PV plugin can keep a set of taken PVs for its own use.

Reply:
Yes, the plugin should keep its own cache, however, what I meant was that I think there needs to be a plugin-level Reserve call so that the plugin can update its cache before more pods are processed.  

Just like today where the scheduler pod cache is updated to account for that pod's resources on the node before it actually gets bound to a node, the volume plugin needs to update its PV cache to mark the PV as taken so that subsequent pods will not try to take the same one.

Reply:
The use-case make sense. I updated the proposal and added an extension point at "Reserve".